src/triage/component/architect/feature_block.py

```diff
@@ -0,0 +1,235 @@
+from abc import ABC, abstractmethod
+
+import logging
+import sqlparse
+from triage.database_reflection import table_exists, table_columns
+from triage.util.db import run_statements
+
+
+class FeatureBlock(ABC):
+    def __init__(
+        self,
+        db_engine,
+        cohort_table,
+        as_of_dates,
+        features_schema_name=None,
+        feature_start_time=None,
+        features_ignore_cohort=False,
+    ):
+        self.db_engine = db_engine
+        self.cohort_table_name = cohort_table
+        self.as_of_dates = as_of_dates
+        self.features_schema_name = features_schema_name
+        self.feature_start_time = feature_start_time
+        self.features_ignore_cohort = features_ignore_cohort
+
+    @property
+    @abstractmethod
+    def final_feature_table_name(self):
+        "The name of the final table with all features filled in (no missing values)"
+        pass
+
+    @property
+    @abstractmethod
+    def feature_columns(self):
+        """
+        The list of feature columns in the final, postimputation table
+
+        Should exclude any index columns (e.g. entity id, date)
+        """
+        pass
+
+    @property
+    @abstractmethod
+    def preinsert_queries(self):
+        """
+        Return all queries that should be run before inserting any data.
+
+        Returns a list of queries/executable statements
+        """
+        pass
+
+    @property
+    @abstractmethod
+    def insert_queries(self):
+        """
+        Return all inserts to populate this data. Each query in this list should be parallelizable.
+
+        Returns a list of queries/executable statements
+        """
+        pass
+
+    @property
+    @abstractmethod
+    def postinsert_queries(self):
+        """
+        Return all queries that should be run after inserting all data
+
+        Returns a list of queries/executable statements
+        """
+        pass
+
+    @property
+    @abstractmethod
+    def imputation_queries(self):
+        """
+        Return all queries that should be run to fill in missing data with imputed values.
+
+        Returns a list of queries/executable statements
+        """
+        pass
+
+    def _cohort_table_sub(self):
+        """Helper function to ensure we only include state table records
+        in our set of input dates and after the feature_start_time.
+        """
+        datestr = ", ".join(["'%s'::date" % dt for dt in self.as_of_dates])
```

Comment:
no need for list, (and can use f-string i would think 🤷‍♂️)

```suggestion
        datestr = ", ".join(f"'{dt}'::date" for dt in self.as_of_dates)
```